File: leetcode/4 Median of Two Sorted Arrays/leetcode4.cpp
#include<vector>
#include<algorithm>
using namespace std;
class Solution {
public:
	double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
		int m = nums1.size();
		int n = nums2.size();
		if (m > n)
		{
			nums1.swap(nums2);
			swap(n, m);
		}
		int imin = 0;
		int imax = m;
		int half_len = (m + n + 1) / 2;
		while (imin <= imax)
		{
			int i = (imin + imax) / 2;
			int j = half_len - i;
			if (j > 0 && i<m && nums2[j - 1]>nums1[i])
				imin = i + 1;
			else if (i > 0 && j<n && nums1[i - 1]>nums2[j])
				imax = i - 1;
			else
			{
				int temp1, temp2;
				if (i == 0)
					temp1 = nums2[j - 1];
				else if (j == 0)
					temp1 = nums1[i - 1];
				else
					temp1 = max(nums1[i - 1], nums2[j - 1]);
				if ((m + n) % 2 == 1)
					return temp1;
				if (i == m)
					temp2 = nums2[j];
				else if (j == n)
					temp2 = nums1[i];
				else
					temp2 = min(nums1[i], nums2[j]);
				return static_cast<double>(temp1 + temp2) / 2;
			}
		}
	}
};
int main()
{
	Solution s;
	vector<int> nums1;
	vector<int> nums2{ 1 };
	auto result=s.findMedianSortedArrays(nums1, nums2);
	return 0;
}

File: leetcode/4 Median of Two Sorted Arrays/leetcode4.py
class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        m, n = len(nums1), len(nums2)
        if m > n:
            nums1, nums2, m, n = nums2, nums1, n, m
        i_min = 0
        i_max = m
        half_len = (m+n+1)//2
        while i_min <= i_max:
            i = (i_min + i_max)//2
            j = half_len - i
            if j > 0 and i < m and nums2[j-1] > nums1[i]:
                i_min = i + 1
            elif i > 0 and j < n and nums1[i-1] > nums2[j]:
                i_max = i -1
            else:
                if i == 0:
                    num1 = nums2[j-1]
                elif j == 0:
                    num1 = nums1[i-1]
                else:
                    num1 = max(nums1[i-1], nums2[j-1])

                if (m+n) % 2 == 1:
                    return num1
                if i==m:
                    num2 = nums2[j]
                elif j == n:
                    num2 = nums1[i]
                else:
                    num2 = min(nums1[i], nums2[j])
                return (num1+num2) / 2.0

s = Solution()
print(s.findMedianSortedArrays([1,3,5,6,7],[2,5,6,7,8,9,11]))